### PR TITLE
[CLOB-919] log errors in clob DeliverTx handlers

### PR DESCRIPTION
**Pull request by jakob-dydx**
- adds error logging for any errors returned from DeliverTx handlers for CLOB msgs. Wraps the errors with block height

### Summary by CodeRabbit

- New Feature: Added a new function `LogErrorWithBlockHeight` in the `errorlib` package to log errors with additional block height information.
- Refactor: Updated various functions across the `clob` keeper to use the new `LogErrorWithBlockHeight` for error logging, improving error visibility and debugging capabilities.
- Test: Added tests for the new `LogErrorWithBlockHeight` function to ensure correct functionality.
- Chore: Modified several function signatures to return named variables for better readability and maintainability.

### CODE REVIEW

File: protocol/lib/error/logging.go [hunk 1]
```diff
@@ -36,3 +40,14 @@ func WrapErrorWithSourceModuleContext(err error, module string) error {
...
+// LogErrorWithBlockHeight logs an error using the cosmos sdk Context's logger, appending the block
+// height to the error message.
+func LogErrorWithBlockHeight(ctx sdk.Context, err error) {
+	err = errorsmod.Wrapf(
+		err,
+		"Block height: %d",
+		ctx.BlockHeight(),
+	)
+	ctx.Logger().Error(err.Error())
+}
```
[the remark this answers is not in the document]
Comment (lucas-dydx) [reply protocol/lib/error/logging.go:53]: As much as I hate to agree with the rabbit, I agree with the rabbit that we should verify `err != nil` before this function is called. Maybe you can add an error log if this is called with a `nil` error?

Also as a suggestion, add callback as well (`CheckTx` / `DeliverTx` / `RecheckTx` / etc)

File: protocol/lib/error/logging.go [hunk 2]
```diff
@@ -36,3 +40,14 @@ func WrapErrorWithSourceModuleContext(err error, module string) error {
...
+
+// LogErrorWithBlockHeight logs an error using the cosmos sdk Context's logger, appending the block
+// height to the error message.
+func LogErrorWithBlockHeight(ctx sdk.Context, err error) {
+	err = errorsmod.Wrapf(
+		err,
+		"Block height: %d",
+		ctx.BlockHeight(),
+	)
+	ctx.Logger().Error(err.Error())
```
Comment (lucas-dydx) [inline protocol/lib/error/logging.go:52]: Nit: accept the logger as an arg so the `module` key is set properly

File: protocol/x/clob/keeper/msg_server_proposed_operations.go [hunk 3]
```diff
@@ -3,28 +3,27 @@ package keeper
...
 	msg *types.MsgProposedOperations,
-) (*types.MsgProposedOperationsResponse, error) {
+) (resp *types.MsgProposedOperationsResponse, err error) {
 	ctx := sdk.UnwrapSDKContext(goCtx)
 
+	defer func() {
+		if err != nil {
+			errorlib.LogErrorWithBlockHeight(ctx, err)
+		}
+	}()
```
Comment (lucas-dydx) [inline protocol/x/clob/keeper/msg_server_proposed_operations.go:21]: Dig this pattern though, I like it

File: protocol/x/clob/keeper/msg_server_update_clob_pair.go [hunk 4]
```diff
@@ -2,17 +2,27 @@ package keeper
...
+
+	defer func() {
+		if err != nil {
+			errorlib.LogErrorWithBlockHeight(ctx, err)
+		}
+	}()
+
 	if !k.Keeper.HasAuthority(msg.Authority) {
 		return nil, errorsmod.Wrapf(
 			govtypes.ErrInvalidSigner,
```
[the remark this answers is not in the document]
Comment (lucas-dydx) [reply protocol/x/clob/keeper/msg_server_update_clob_pair.go:28]: Please resolve this comment out of shame for how wrong you are

File: protocol/lib/error/logging.go [hunk 5]
```diff
@@ -36,3 +40,14 @@ func WrapErrorWithSourceModuleContext(err error, module string) error {
 	return NewErrorWithLogContext(err).
 		WithLogKeyValue(SourceModuleKey, fmt.Sprintf("x/%v", module))
 }
+
+// LogErrorWithBlockHeight logs an error using the cosmos sdk Context's logger, appending the block
+// height to the error message.
+func LogErrorWithBlockHeight(ctx sdk.Context, err error) {
```
Comment (lucas-dydx) [inline protocol/lib/error/logging.go:46]: Any way to test this function?

Specifically a test that verifies it doesn't panic when provided a nil `err` or something

**Review by lucas-dydx — APPROVED** [review]

File: protocol/lib/error/logging.go [hunk 6]
```diff
@@ -36,3 +39,18 @@ func WrapErrorWithSourceModuleContext(err error, module string) error {
...
+	if err != nil {
+		err = errorsmod.Wrapf(
+			err,
+			"Block height: %d, Callback: %s",
+			blockHeight,
+			callback,
+		)
+		logger.Error(err.Error())
+	} else {
+		logger.Error("LogErrorWithBlockHeight called with nil error")
```
Comment (lucas-dydx) [inline protocol/lib/error/logging.go:54]: Should we add a stack trace here? Up to you

File: protocol/x/clob/keeper/msg_server_create_clob_pair.go [hunk 7]
```diff
@@ -14,7 +17,15 @@ import (
 func (k msgServer) CreateClobPair(
 	goCtx context.Context,
 	msg *types.MsgCreateClobPair,
-) (*types.MsgCreateClobPairResponse, error) {
+) (resp *types.MsgCreateClobPairResponse, err error) {
+	ctx := sdk.UnwrapSDKContext(goCtx)
+
+	defer func() {
+		if err != nil {
+			errorlib.LogErrorWithBlockHeight(ctx.Logger(), err, ctx.BlockHeight(), metrics.DeliverTx)
```
Comment (lucas-dydx) [inline protocol/x/clob/keeper/msg_server_create_clob_pair.go:25]: Can you pass the keepers logger? Specifically using [this](https://github.com/dydxprotocol/v4-chain/blob/46ce587c6a713c469050e462cea7b4322b082b2d/protocol/x/clob/keeper/keeper.go#L133).

Comment (jakob-dydx) [reply protocol/x/clob/keeper/msg_server_create_clob_pair.go:25]: ah I see, my bad. Will update

**Pull request merged**